src/VisualStudio/Core/Impl/CodeModel/Collections/PartialTypeCollection.cs

```diff
@@ -45,25 +41,43 @@ private AbstractCodeType ParentType
 
         private ImmutableArray<EnvDTE.CodeElement> GetParts()
         {
-            var partsBuilder = ImmutableArray.CreateBuilder<EnvDTE.CodeElement>();
-            var symbol = ParentType.LookupSymbol();
-
-            foreach (var location in symbol.Locations.Where(l => l.IsInSource))
+            // Retrieving the parts is potentially very expensive because it can force multiple FileCodeModels to be instantiated.
+            // Here, we cache the result to avoid having to perform these calculations each time GetParts() is called.
+            // This *could* be an issue because it means that a PartialTypeCollection will not necessarily reflect the
+            // current state of the user's code. However, because a new PartialTypeCollection is created everytime a the Parts
+            // property is accessed on CodeClass, CodeStruct or CodeInterface, consumers would hit this behavior rarely.
+            if (this._parts == null)
             {
-                var tree = location.SourceTree;
-                var document = this.Workspace.CurrentSolution.GetDocument(tree);
+                var partsBuilder = ImmutableArray.CreateBuilder<EnvDTE.CodeElement>();
 
-                var fileCodeModelObject = this.Workspace.GetFileCodeModel(document.Id);
-                var fileCodeModel = ComAggregate.GetManagedObject<FileCodeModel>(fileCodeModelObject);
+                var solution = this.Workspace.CurrentSolution;
+                var symbol = ParentType.LookupSymbol();
 
-                var element = fileCodeModel.CodeElementFromPosition(location.SourceSpan.Start, ParentType.Kind);
-                if (element != null)
+                foreach (var location in symbol.Locations.Where(l => l.IsInSource))
                 {
-                    partsBuilder.Add(element);
+                    var tree = location.SourceTree;
+                    var document = solution.GetDocument(tree);
+
+                    if (document != null)
+                    {
+                        var fileCodeModelObject = this.Workspace.GetFileCodeModel(document.Id);
```

Comment:
Yes, I came to the same conclusion, which is what motivated this hardening. I can imagine a couple of situations where this might happen:

* A document containing part of a partial type (say, a .g.i.cs flie) is removed from the project but that change has not yet propagated to the workspace. So, the workspace's current solution still has the document but ```RoslynVisualStudioWorkspace.GetFileCodeModel()``` calls directly into the project system and the file is no longer there.
* A document containing part of a partial type is removed between asking for the document and asking for it's file code model (i.e. a race condition).